cpp/src/arrow/dlpack.cc

```diff
@@ -0,0 +1,97 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/dlpack.h"
+
+#include "arrow/array/array_base.h"
+#include "arrow/dlpack_structure.h"
+#include "arrow/type.h"
+
+namespace arrow {
+
+DLDataType getDLDataType(const Array& arr) {
+  DLDataType dtype;
+  dtype.lanes = 1;
+  dtype.bits = arr.type()->bit_width();
+  switch (arr.type()->id()) {
+    case Type::INT8:
+    case Type::INT16:
+    case Type::INT32:
+    case Type::INT64:
+      dtype.code = DLDataTypeCode::kDLInt;
+      break;
+    case Type::UINT8:
+    case Type::UINT16:
+    case Type::UINT32:
+    case Type::UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case Type::HALF_FLOAT:
+    case Type::FLOAT:
+    case Type::DOUBLE:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case Type::BOOL:
+      dtype.code = DLDataTypeCode::kDLBool;
+      break;
+    default:
+      // TODO
+      break;
+  }
+  return dtype;
+}
+
+struct DLMTensorCtx {
+  std::shared_ptr<ArrayData> ref;
+  std::vector<int64_t> shape;
+  DLManagedTensor tensor;
+};
+
+static void deleter(DLManagedTensor* arg) {
+  delete static_cast<DLMTensorCtx*>(arg->manager_ctx);
+}
+
+DLManagedTensor* toDLPack(const Array& arr) {
+  std::shared_ptr<ArrayData> array_ref = arr.data();
```

Comment:
Does this make a deep copy of the underlying data?

Reply:
I thought not, but am not really sure. AFAIU this is only needed for a reference so I can just as well use `Slice` on an `Array` which for sure doesn't do any copies.

Reply:
Hm, I think I remember now why I chose `.data()` and not `.Slice()`. When accessing the buffer address of the data for `dl_tensor.data` pointer we also go through `array.data()` ([data_](https://github.com/apache/arrow/blob/5306b32370b4c5f5137defe90516a280d9c71796/cpp/src/arrow/array/array_base.h#L254)) so it looked to me like `ArrayData` is an underlying `Array` structure and it would be safe to use it as a reference.

@pitrou what do you think?